### PR TITLE
Add a callback to $tw.boot.boot()

### PR DESCRIPTION
**Pull request by Arlen22**
I am proposing to add a callback to $tw.boot.boot() that would get called once all the startup modules are loaded. I am thinking that this should get it to the point where everything is initialized and ready for action. 

The use case is that I need a way to know when a TiddlyWiki is done loading so I can execute the server command module and give it a request to handle.

### CODE REVIEW

**Comment by Jermolene** [comment]
> The use case is that I need a way to know when a TiddlyWiki is done loading so I can execute the server command module and give it a request to handle.

The usual way of doing that would be by assigning an array of commands to `$tw.boot.argv` before calling `$tw.boot.boot()`, as you see in `tiddlywiki.js` in the root of the repo. Would that work for you?

**Comment by Arlen22** [comment]
Well, the code I am working with is loading the TiddlyWiki object as tiddlywiki.js does, but then it needs to know when the loading is complete, as some modules are asynchronous. I wish the whole startup was asynchronous :)

```ts
function loadTiddlyWiki(prefix, folder) {
    return new Promise(resolve => {
        const $tw = require("tiddlywiki/boot/boot.js").TiddlyWiki();
        $tw.boot.argv = [folder];
        const execute = $tw.boot.executeNextStartupTask;
        $tw.boot.executeNextStartupTask = function () {
            const res = execute();
            if (res === false) complete();
        }
        function complete() {
            //we use $tw.modules.execute so that the module has its respective $tw variable.
            var serverCommand = $tw.modules.execute('$:/core/modules/commands/server.js').Command;
            var command = new serverCommand([], { wiki: $tw.wiki });
            var server = command.server;

            server.set({
                rootTiddler: "$:/core/save/all",
                renderType: "text/plain",
                serveType: "text/html",
                username: "",
                password: "",
                pathprefix: prefix
            });

            loadedFolders[prefix] = {
                $tw,
                prefix,
                folder,
                server,
                handler: server.requestHandler.bind(server)
            }
            resolve(loadedFolders[prefix].handler);
        }
        $tw.boot.boot();
        $tw.wiki.addTiddler({
            "text": "$protocol$//$host$" + prefix + "/",
            "title": "$:/config/tiddlyweb/host"
        });
    })
};
```

**Comment by Arlen22** [comment]
As you can see, I currently just override executeNextStartupTask and wait for it to return false, then call complete.

**Comment by Jermolene** [comment]
Thanks @Arlen22. I'm sure I'm missing something obvious, but could you not do something like this:

```
function loadTiddlyWiki(prefix, folder) {
    return new Promise(resolve => {
        const $tw = require("tiddlywiki/boot/boot.js").TiddlyWiki();
        $tw.boot.argv = [
            folder,
            "--server","8080","$:/core/save/all","text/plain","text/html","","",host,prefix
        ];
        $tw.boot.boot();
        $tw.wiki.addTiddler({
            "text": "$protocol$//$host$" + prefix + "/",
            "title": "$:/config/tiddlyweb/host"
        });
    })
};
```

**Comment by Arlen22** [comment]
Oh, I follow you. No, what I'm actually doing is loading multiple TiddlyWiki instances into the same Node instance and then calling the correct one based on the prefix received. It is basically what I have done before, but this time the first request loads TiddlyWiki instead of loading TiddlyWiki at startup.

https://gist.github.com/Arlen22/bbd852f68e328165e49f

**Comment by Jermolene** [comment]
Hi @Arlen22 aha, that makes sense, interesting stuff.

Looking at the code, it seems you are catering for `$tw.boot.startup` to be called with `(callback)` as well as `(options, callback)` and the existing `(options)`. Is it worth supporting `(callback)`? The existing function only supports `(options)`; adding the callback parameter looks like a backwards compatible extension of the old function signature.

**Comment by Arlen22** [comment]
Actually, I didn't know that startup ever had a callback. I just did that so options would be optional (callback is made optional with (`callback && callback()`). Should we make callback a property on the options object? I would have no objection to that.

**Comment by Jermolene** [comment]
Hi @Arlen22 apologies for the confusion.

> Should we make callback a property on the options object? I would have no objection to that.

Yes please.

**Comment by Arlen22** [comment]
Ok, done. How does that look.

**Comment by danielo515** [comment]
This has been requested a couple of times before. We should link the issues

**Comment by danielo515** [comment]
The issue I was talking about is #2766

File: boot/boot.js [hunk 1]
```diff
@@ -2022,7 +2017,7 @@ $tw.boot.startup = function(options, callback) {
 	$tw.boot.executedStartupModules = Object.create(null);
 	$tw.boot.disabledStartupModules = $tw.boot.disabledStartupModules || [];
 	// Repeatedly execute the next eligible task
-	$tw.boot.executeNextStartupTask(callback);
+	$tw.boot.executeNextStartupTask(options.callback);
```
Comment (sukima) [inline boot/boot.js]: Is this null safe? `$tw.boot.executeNextStartupTask(undefined)`? What happens when `options.callback` is not defined or is not a function?

File: boot/boot.js [hunk 2]
```diff
@@ -2022,7 +2017,7 @@ $tw.boot.startup = function(options, callback) {
...
 };
 
 /*
```
Comment (sukima) [inline boot/boot.js]: This commit could be amended to the previous one.

File: boot/boot.js [hunk 3]
```diff
@@ -2167,7 +2168,7 @@ $tw.boot.boot = function() {
 	// Preload any encrypted tiddlers
 	$tw.boot.decryptEncryptedTiddlers(function() {
 		// Startup
-		$tw.boot.startup();
+		$tw.boot.startup(callback);
```
Comment (sukima) [inline boot/boot.js]: I thought `$tw.boot.startup()` now takes an `options` object not a `callback` function.

File: boot/boot.js [hunk 4]
```diff
@@ -2064,14 +2064,15 @@ $tw.boot.executeNextStartupTask = function() {
...
-				return $tw.boot.executeNextStartupTask();
+				return $tw.boot.executeNextStartupTask(callback);
 			} else {
 				task.startup(asyncTaskCallback);
 				return true;
 			}
 		}
 		taskIndex++;
 	}
+	if(typeof callback === 'function') callback();
```
Comment (sukima) [inline boot/boot.js]: Ahh so it is null safe. Cool.

File: boot/boot.js (hunk 3 above)
Comment (Arlen22) [inline boot/boot.js]: Good catch. Fixed.

**Comment by danielo515** [comment]
The only thing I'm missing is the documentation section under the dev edition.
Just to be clear, the intention is that the provided callback will be executed when tiddlywiki has fully boot, right?

**Comment by Arlen22** [comment]
Technically when the last startup task has completed (asynchronously, if the task is async).

**Comment by danielo515** [comment]
Exactly what I needed

**Comment by Jermolene** [comment]
Thanks @Arlen22

**Pull request merged**